### PR TITLE
Sticky Selection

### PR DESCRIPTION
**Pull request by Templarfreak**
Refine micro mechanics. Allow for more organic feeling controls.

This mechanic is very helpful for a number of micro techniques, such as:

Splitting.
Concaves.
And Kiting.

Mechanics that allow for players to show off their skill in unit control
makes a more exciting to watch and more rewarding to play game without raising the skill floor.

Video showcasing the change in work:

http://www.mediafire.com/watch/dj9i9bba4r4inbt/OpenRA.Game_2015-05-17_03-20-27-896.avi

### CODE REVIEW

**Comment by Templarfreak** [comment]
Er, to clarify what this does is if someone makes a drag select box then it won't deselect units that are currently in the player's selection if there are no units to select in that drag select box.

**Comment by x-a-n-a-x** [comment]
a GIF or small video would be cool to clearly see what it changes.

**Comment by abcdefg30** [comment]
Could you please use tabs instead of spaces? Thanks.

**Comment by Templarfreak** [comment]
Spaces happened because I copy/pasted it over from another file. I used tabs originally but it turned them into spaces. xD

Will make a video.

**Comment by Templarfreak** [comment]
Here is a video: http://www.mediafire.com/watch/dj9i9bba4r4inbt/OpenRA.Game_2015-05-17_03-20-27-896.avi

**Comment by WolfGaming** [comment]
If you could make it an option that you can switch between the two behaviors I would be very happy since I loath playing with sticky selection xD

**Comment by Templarfreak** [comment]
I just finished making it so Sticky Selection is only on when Left-Click orders is on. For a proper toggle, I'll have to spend tomorrow figuring out how the UI works for that.

**Comment by Templarfreak** [comment]
I have no idea how to "Squish" all those changes into one...

**Comment by Unit158** [comment]
```bash
git fetch upstream bleed # Or "openra bleed", whichever you use.
git rebase -i FETCH_HEAD
#now, choose the commits you'd like to squash
```

**Comment by Templarfreak** [comment]
For both of those it says it's not a git repository.

**Comment by Unit158** [comment]
You'll need to set OpenRA/OpenRA as a remote then. Come on the IRC for help.

**Comment by dan9550** [comment]
What @WolfGaming said if this was by default you would make me a sad panda lol

Toggle-able in game would be my preference.

**Comment by Mailaender** [comment]
This seems to be very useful in terms of improving usability. However we can't accept this GitHub web interface broken commit spam. https://help.github.com/articles/using-git-rebase/ has a tutorial, but as this is often confusing for beginners, getting talked through the process in the chat may be necessary. See http://www.openra.net/community/ on how to reach us.

**Comment by Templarfreak** [comment]
Changes requested made. (I think)

**Comment by Templarfreak** [comment]
Changed comment to look not dumb.

File: OpenRA.Game/Selection.cs [hunk 1]
```diff
@@ -33,40 +33,65 @@ public bool Contains(Actor a)
...
+            if (voicedUnit != null)
+                Sound.PlayVoice("Select", voicedUnit, voicedUnit.Owner.Country.Race);
+                                           }
+
+            foreach (var a in newSelection)
+                foreach (var sel in a.TraitsImplementing<INotifySelected>())
+                    sel.Selected(a);
+            foreach (var ns in world.WorldActor.TraitsImplementing<INotifySelection>())
+                ns.SelectionChanged();
+        }
```
Comment (Mailaender) [inline OpenRA.Game/Selection.cs]: Can you replace the space indention with tabs please. We provide an @EditorConfig that may be of help to you.

**Comment by Templarfreak** [comment]
I don't know how to use that... What exactly do I do to compile it? I don't see a project file.

File: OpenRA.Game/Selection.cs [hunk 2]
```diff
@@ -33,40 +33,65 @@ public bool Contains(Actor a)
...
+                        actors.UnionWith(newSelection);
+               }
+            }
+
+            var voicedUnit = actors.FirstOrDefault(a => a.Owner == world.LocalPlayer && a.IsInWorld && a.HasVoices());
+            if (newSelection.Count() != 0)
+            {
+            if (voicedUnit != null)
+                Sound.PlayVoice("Select", voicedUnit, voicedUnit.Owner.Country.Race);
+                                           }
```
Comment (Mailaender) [inline OpenRA.Game/Selection.cs]: Something went wrong here, too with indentions.

**Comment by Mailaender** [comment]
See https://msdn.microsoft.com/en-us/library/ms165330%28v=vs.90%29.aspx?f=255&MSPPError=-2147217396

**Comment by Mailaender** [comment]
See http://edglogowski.azurewebsites.net/?p=10841

File: OpenRA.Game/Selection.cs [hunk 3]
```diff
@@ -33,40 +33,65 @@ public bool Contains(Actor a)
 			return actors.Contains(a);
 		}
 
-		public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine, bool isClick)
+	public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine, bool isClick)
```
Comment (Mailaender) [inline OpenRA.Game/Selection.cs]: Everything is off by one tab/indention, but the spaces are gone.

File: OpenRA.Game/Selection.cs [hunk 4]
```diff
@@ -35,37 +35,62 @@ public bool Contains(Actor a)
...
+				actors.Clear();
+				actors.UnionWith(newSelection);
+			}
 			}
 			else
 			{
-				if (isCombine)
+			if (isCombine)
+				actors.UnionWith(newSelection);
+			 else
```
Comment (Mailaender) [inline OpenRA.Game/Selection.cs]: spurious space

File: OpenRA.Game/Selection.cs [hunk 5]
```diff
@@ -35,37 +35,62 @@ public bool Contains(Actor a)
 
 		public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine, bool isClick)
 		{
-			if (isClick)
+		var classicmouse = Game.Settings.Game.UseClassicMouseStyle;
```
Comment (Mailaender) [inline OpenRA.Game/Selection.cs]: Still missing one indention step.

File: OpenRA.Game/Selection.cs [hunk 6]
```diff
@@ -35,37 +35,62 @@ public bool Contains(Actor a)
...
+			if (isCombine)
+				actors.SymmetricExceptWith(adjNewSelection);
+			 else
 			{
-				var adjNewSelection = newSelection.Take(1);	/* TODO: select BEST, not FIRST */
-				if (isCombine)
-					actors.SymmetricExceptWith(adjNewSelection);
-				else
+				if (classicmouse == true)
+			{
```
Comment (Mailaender) [inline OpenRA.Game/Selection.cs]: This is off by two indentions.

**Comment by Templarfreak** [comment]
Okay, if all the indentation's right now I'll go ahead and squash it all again.

**Comment by Templarfreak** [comment]
Okay, gonna squish 'em now.

**Comment by obrakmann** [comment]
* A few codestyle issues. At a first glance, the indentation is wrong in some lines.
* The checkbox will need to be added to the cnc mod as well
* You never actually use Settings.StickySelection 
* We'll have to figure out the best place for the checkbox to live in the UI, because it's so cramped in there and everything collides

File: OpenRA.Game/Selection.cs [hunk 7]
```diff
@@ -42,8 +43,19 @@ public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine
...
+							actors.UnionWith(adjNewSelection);
+						}
+						else
+						actors.Clear();
+						actors.UnionWith(newSelection);
+					}
+					else
 					actors.Clear();
-					actors.UnionWith(adjNewSelection);
+					actors.UnionWith(newSelection);
```
Comment (obrakmann) [inline OpenRA.Game/Selection.cs]: This block needs one more level of indentation, and it has to be wrapped in braces, too.

File: OpenRA.Game/Selection.cs [hunk 8]
```diff
@@ -52,14 +64,26 @@ public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine
 					actors.UnionWith(newSelection);
 				else
 				{
+					if (stickyselect == true)
+					{
+						if (newSelection.Count() == 0)
+							actors.UnionWith(newSelection);
+						else
+						actors.Clear();
+						actors.UnionWith(newSelection);
```
Comment (obrakmann) [inline OpenRA.Game/Selection.cs]: Here as well: one more level of indentation, and braces are missing on the else block

File: OpenRA.Game/Selection.cs [hunk 9]
```diff
@@ -52,14 +64,26 @@ public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine
...
+					}
+					else
 					actors.Clear();
 					actors.UnionWith(newSelection);
```
Comment (obrakmann) [inline OpenRA.Game/Selection.cs]: and again

**Comment by Templarfreak** [comment]
I really have no idea how many times I've set those indentations and VS undid them...

**Comment by WolfGaming** [comment]
You probably havent set it up to keep tabs or something.

Try going to Options/Text Editor/C#/Tabs and check if the tab/indent size are both set to 4 and Keep tabs is selected.

**Comment by abcdefg30** [comment]
Needs a rebase.

**Comment by Templarfreak** [comment]
Will need to come back to this later, lots of stuff just came up. Sorry, guys! (Don't worry, I'm not disappearing. I can come back to this in a couple weeks)

**Comment by pchote** [comment]
Sure thing @Templarfreak.  I'll close this pr for now, and you can either either reopen it if/when you decide to continue on this version, or open a new one if you decide to do it over.

**Pull request closed without merging**